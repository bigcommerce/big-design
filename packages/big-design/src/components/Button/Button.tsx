import React, { ComponentPropsWithoutRef, forwardRef, memo, Ref } from 'react';

import { MarginProps } from '../../helpers';
import { ProgressCircle } from '../ProgressCircle';

import { ContentWrapper, LoadingSpinnerWrapper, StyledButton } from './styled';

interface PrivateProps {
  forwardedRef: Ref<HTMLButtonElement>;
}

export interface ButtonProps extends ComponentPropsWithoutRef<'button'>, MarginProps {
  readonly actionType?: 'normal' | 'destructive';
  readonly children?: React.ReactNode;
  readonly iconLeft?: React.ReactNode;
  readonly iconOnly?: React.ReactNode;
  readonly iconRight?: React.ReactNode;
  readonly isLoading?: boolean;
  readonly mobileWidth?: 'auto' | '100%';
  readonly variant?: 'primary' | 'secondary' | 'subtle' | 'utility';
}

const LoadingSpinner = () => (
  <LoadingSpinnerWrapper alignItems="center">
    <ProgressCircle size="xxSmall" />
  </LoadingSpinnerWrapper>
);

const RawButton: React.FC<ButtonProps & PrivateProps> = memo(
  ({ forwardedRef, isLoading, disabled, ...props }) => {
    return (
      <StyledButton
        className="bd-button"
        {...props}
        disabled={isLoading || disabled}
        ref={forwardedRef}
      >
        {isLoading ? <LoadingSpinner /> : null}
        <ContentWrapper isLoading={isLoading}>
          {!props.iconOnly && props.iconLeft}
          {props.iconOnly}
          {!props.iconOnly && props.children}
          {!props.iconOnly && props.iconRight}
        </ContentWrapper>
      </StyledButton>
    );
  },
);

export const StyleableButton = forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => (
  <RawButton {...props} forwardedRef={ref} />
));

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, style, ...props }, ref) => <RawButton {...props} forwardedRef={ref} />,
);

const defaultProps = {
  actionType: 'normal' as const,
  isLoading: false,
  mobileWidth: '100%' as const,
  variant: 'primary' as const,
};

Button.displayName = 'Button';
Button.defaultProps = { ...defaultProps };

StyleableButton.displayName = 'StyleableButton';
StyleableButton.defaultProps = { ...defaultProps };
