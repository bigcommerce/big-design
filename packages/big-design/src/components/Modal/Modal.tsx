import { CloseIcon } from '@bigcommerce/big-design-icons';
import { createFocusTrap, FocusTrap } from 'focus-trap';
import React, { useEffect, useId, useMemo, useRef, useState } from 'react';
import { createPortal } from 'react-dom';

import { typedMemo } from '../../utils';
import { Button, ButtonProps } from '../Button';
import { MessagingButton } from '../Button/private';
import { H2 } from '../Typography';

import {
  StyledModal,
  StyledModalActions,
  StyledModalBody,
  StyledModalClose,
  StyledModalContent,
  StyledModalHeader,
} from './styled';

export interface ModalProps {
  readonly actions?: ModalAction[];
  readonly backdrop?: boolean;
  readonly children?: React.ReactNode;
  readonly closeOnClickOutside?: boolean;
  readonly closeOnEscKey?: boolean;
  readonly header?: string;
  readonly isOpen?: boolean;
  readonly variant?: 'modal' | 'dialog';
  onClose?(): void;
}

export interface ModalAction extends Omit<ButtonProps, 'children'> {
  text?: string;
}

export const Modal: React.FC<ModalProps> = typedMemo((props) => {
  const [modalContainer, setModalContainer] = useState<HTMLDivElement | null>(null);

  useEffect(() => {
    const container = document.createElement('div');

    document.body.appendChild(container);
    setModalContainer(container);
  }, []);

  useEffect(() => {
    return () => {
      if (modalContainer) {
        document.body.removeChild(modalContainer);
      }
    };
  }, [modalContainer]);

  return props.isOpen && modalContainer
    ? createPortal(<InternalModal {...props} />, modalContainer)
    : null;
});

const InternalModal: React.FC<ModalProps> = ({
  actions,
  backdrop = true,
  children,
  closeOnClickOutside = false,
  closeOnEscKey = true,
  header,
  onClose = () => null,
  variant = 'modal',
}) => {
  const initialBodyOverflowYRef = useRef('');
  const internalTrap = useRef<FocusTrap | null>(null);
  const headerUniqueId = useId();
  const [modalRef, setModalRef] = useState<HTMLDivElement | null>(null);

  const onClickAway = (event: React.MouseEvent) => {
    if (closeOnClickOutside && modalRef === event.target) {
      onClose();
    }
  };

  const onKeyDown = (event: React.KeyboardEvent) => {
    if (closeOnEscKey && event.key === 'Escape') {
      onClose();
    }
  };

  // Disable scroll on body when modal is open
  useEffect(() => {
    initialBodyOverflowYRef.current = document.body.style.overflowY || '';
    document.body.style.overflowY = 'hidden';

    return () => {
      document.body.style.overflowY = initialBodyOverflowYRef.current;
    };
  }, []);

  // Setup focus-trap
  useEffect(() => {
    if (modalRef && internalTrap.current === null) {
      internalTrap.current = createFocusTrap(modalRef, { fallbackFocus: modalRef });
      internalTrap.current.activate();
    }

    return () => {
      internalTrap.current?.deactivate();
    };
  }, [internalTrap, modalRef]);

  const renderedClose = useMemo(
    () =>
      variant === 'modal' && (
        <StyledModalClose>
          <MessagingButton iconOnly={<CloseIcon title="Close" />} onClick={onClose} />
        </StyledModalClose>
      ),
    [onClose, variant],
  );

  const renderedHeader = useMemo(
    () =>
      header &&
      typeof header === 'string' && (
        <StyledModalHeader id={headerUniqueId}>
          <H2 margin="none">{header}</H2>
        </StyledModalHeader>
      ),
    [header, headerUniqueId],
  );

  const renderedActions = useMemo(
    () =>
      Array.isArray(actions) && (
        <StyledModalActions justifyContent="flex-end">
          {actions.map(({ text, onClick, ...props }, index) => (
            <Button
              key={index}
              {...props}
              onClick={(event) => {
                internalTrap.current?.deactivate();

                if (typeof onClick === 'function') {
                  onClick(event);
                }
              }}
            >
              {text}
            </Button>
          ))}
        </StyledModalActions>
      ),
    [actions],
  );

  return (
    <StyledModal
      backdrop={backdrop}
      onClick={onClickAway}
      onKeyDown={onKeyDown}
      ref={setModalRef}
      tabIndex={-1}
      variant={variant}
    >
      <StyledModalContent aria-labelledby={headerUniqueId} flexDirection="column" variant={variant}>
        {renderedClose}
        {renderedHeader}
        <StyledModalBody>{children}</StyledModalBody>
        {renderedActions}
      </StyledModalContent>
    </StyledModal>
  );
};
