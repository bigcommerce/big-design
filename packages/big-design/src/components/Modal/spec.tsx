import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';

import 'jest-styled-components';

import { Button } from '../Button';
import { Text } from '../Typography';

import { Modal } from './Modal';

test('render open modal', () => {
  const text = 'This is a modal';
  render(<Modal isOpen={true}>{text}</Modal>);

  expect(document.body).toMatchSnapshot();
  expect(screen.getByText(text)).toBeInTheDocument();
});

test('render open modal without backdrop', () => {
  const text = 'This is a modal';
  const { queryByText } = render(
    <Modal backdrop={false} isOpen={true}>
      {text}
    </Modal>,
  );

  expect(document.body).toMatchSnapshot();
  expect(screen.getByText(text)).toBeInTheDocument();
});

test('render closed modal', () => {
  const text = 'This is a modal';

  render(<Modal isOpen={false}>{text}</Modal>);

  expect(document.body).toMatchSnapshot();
  expect(screen.queryByText(text)).not.toBeInTheDocument();
});

test('open/hides when props changes', () => {
  const text = 'This is a modal';
  const { queryByText, rerender } = render(<Modal isOpen={false}>{text}</Modal>);

  expect(screen.queryByText(text)).not.toBeInTheDocument();

  rerender(<Modal isOpen={true}>{text}</Modal>);

  expect(screen.getByText(text)).toBeInTheDocument();
});

test('triggers onClose when pressing esc', async () => {
  const text = 'This is a modal';
  const onClose = jest.fn();

  render(
    <Modal isOpen={true} onClose={onClose}>
      {text}
    </Modal>,
  );

  const element = await screen.findByText(text);

  fireEvent.keyDown(element, { key: 'Escape' });

  expect(onClose).toHaveBeenCalled();
});

test('does not trigger onClose when pressing esc and closeOnEscKey is false', async () => {
  const text = 'This is a modal';
  const onClose = jest.fn();

  render(
    <Modal closeOnEscKey={false} isOpen={true} onClose={onClose}>
      {text}
    </Modal>,
  );

  const element = await screen.findByText(text);

  fireEvent.keyDown(element, { key: 'Escape' });

  expect(onClose).not.toHaveBeenCalled();
});

test('trigger onClose when clicking outside the modal', async () => {
  const text = 'This is a modal';
  const onClose = jest.fn();

  render(
    <Modal closeOnClickOutside={true} isOpen={true} onClose={onClose}>
      {text}
    </Modal>,
  );

  const element = await screen.findByRole('dialog');

  fireEvent.click(element);

  expect(onClose).toHaveBeenCalled();
});

test('do not trigger onClose when clicking outside the modal and closeOnClickOutside is false', async () => {
  const text = 'This is a modal';
  const onClose = jest.fn();

  render(
    <Modal isOpen={true} onClose={onClose}>
      {text}
    </Modal>,
  );

  const element = await screen.findByRole('dialog');

  fireEvent.click(element);

  expect(onClose).not.toHaveBeenCalled();
});

test('do not trigger onClose when clicking inside the modal', async () => {
  const onClose = jest.fn();

  render(
    <Modal isOpen={true} onClose={onClose}>
      <p data-testid="inside-modal">Content</p>
    </Modal>,
  );

  const element = await screen.findByTestId('inside-modal');

  fireEvent.click(element);

  expect(onClose).not.toHaveBeenCalled();
});

test('render close button on modal variation', () => {
  const text = 'This is a modal';
  const onClose = jest.fn();

  const { queryByTitle } = render(
    <Modal isOpen={true} onClose={onClose} variant="modal">
      {text}
    </Modal>,
  );

  expect(screen.getByTitle('Close')).toBeInTheDocument();
});

test('do not render close button on dialog variation', () => {
  const text = 'This is a modal';
  const onClose = jest.fn();

  const { queryByTitle, queryByText } = render(
    <Modal isOpen={true} onClose={onClose} variant="dialog">
      {text}
    </Modal>,
  );

  expect(screen.getByText(text)).toBeInTheDocument();
  expect(queryByTitle('Close')).not.toBeInTheDocument();
});

test('do not pull focus to open modal that is rerendered', async () => {
  const text = 'This is a modal';
  const { queryByText, queryByRole, rerender } = render(
    <Modal isOpen={false}>
      {text}
      <input id="focusTest" />
    </Modal>,
  );

  // Modal not opened
  expect(screen.queryByText(text)).not.toBeInTheDocument();

  // Modal Opened
  rerender(
    <Modal isOpen={true}>
      {text}
      <input id="focusTest" />
    </Modal>,
  );

  // Expect Modal to have focus
  expect(screen.getByText(text)).toBeInTheDocument();

  await waitFor(() => expect(screen.queryByRole('dialog')).toHaveFocus());

  const input = document.getElementById('focusTest');

  // Focus on input
  if (input !== null) {
    input.focus();

    expect(input).toHaveFocus();

    // Rerender as open
    rerender(
      <Modal isOpen={true}>
        {text}
        <input id="focusTest" />
      </Modal>,
    );

    // Expect input to still have focus and not modal
    expect(input).toHaveFocus();
    expect(screen.queryByRole('dialog')).not.toHaveFocus();
  }
});

test('body has scroll locked on modal open', () => {
  const { rerender } = render(<Modal isOpen={false} />);

  expect(document.body).toHaveStyle({ overflowY: '' });

  rerender(<Modal isOpen={true} />);

  expect(document.body).toHaveStyle({ overflowY: 'hidden' });

  rerender(<Modal isOpen={false} />);

  expect(document.body).toHaveStyle({ overflowY: '' });
});

test('renders header', () => {
  const { getByText } = render(<Modal header="Header Title" isOpen={true} />);

  expect(screen.getByText('Header Title')).toBeInTheDocument();
});

test('header ignores components', () => {
  // @ts-expect-error ignoring since header={Component} is not a valid prop
  const { container } = render(<Modal header={<Text>Header Title</Text>} isOpen={true} />);

  expect(container.querySelector('h2')).toBeNull();
});

test('renders actions', () => {
  const { getAllByRole } = render(
    <Modal actions={[{ text: 'Cancel' }, { text: 'Apply' }]} isOpen={true} />,
  );

  expect(getAllByRole('button')).toHaveLength(3);
});

test('action button triggers onClick', () => {
  const onClick = jest.fn();

  const { getAllByRole } = render(<Modal actions={[{ text: 'Apply', onClick }]} isOpen={true} />);
  const button = getAllByRole('button')[1];

  fireEvent.click(button);

  expect(onClick).toHaveBeenCalled();
});

test('renders secondary action button', () => {
  const { getAllByRole } = render(
    <Modal actions={[{ text: 'Apply', variant: 'secondary' }]} isOpen={true} />,
  );
  const button = getAllByRole('button')[1];

  expect(button).toMatchSnapshot();
});

test('renders destructive action button', () => {
  const { getAllByRole } = render(
    <Modal actions={[{ text: 'Apply', actionType: 'destructive' }]} isOpen={true} />,
  );
  const button = getAllByRole('button')[1];

  expect(button).toMatchSnapshot();
});

test('unmounts appropriately', () => {
  const onClick = jest.fn();
  const { getByTestId, unmount } = render(<Modal isOpen={true} />);

  unmount();

  render(
    <Button data-testid="button" onClick={onClick}>
      Test
    </Button>,
  );

  const button = getByTestId('button');

  button.click();

  // Make sure mouse events still work for other components
  expect(onClick).toHaveBeenCalledTimes(1);
});

test('body overflowY should reset on unmount', () => {
  const { unmount } = render(<Modal isOpen={true} />);

  expect(document.body).toHaveStyle({ overflowY: 'hidden' });

  unmount();

  expect(document.body).toHaveStyle({ overflowY: '' });
});
