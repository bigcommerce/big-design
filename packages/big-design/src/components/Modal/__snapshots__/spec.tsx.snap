// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`render closed modal 1`] = `
<body
  style=""
>
  <div />
  <div />
</body>
`;

exports[`render open modal 1`] = `
.c8 {
  vertical-align: middle;
  height: 1.5rem;
  width: 1.5rem;
}

.c1 {
  box-sizing: border-box;
}

.c2 {
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  -webkit-transition: all 150ms ease-out;
  transition: all 150ms ease-out;
  -webkit-transition-property: background-color,border-color,box-shadow,color;
  transition-property: background-color,border-color,box-shadow,color;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 1px solid #D9DCE9;
  border-radius: 0.25rem;
  color: #FFFFFF;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  font-size: 1rem;
  font-weight: 400;
  height: 2.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 2rem;
  outline: none;
  padding: 0 1rem;
  position: relative;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  width: 100%;
  background-color: #3C64F4;
  border-color: #3C64F4;
  font-weight: 600;
}

.c5:focus {
  outline: none;
}

.c5[disabled] {
  border-color: #D9DCE9;
  pointer-events: none;
}

.c5 + .bd-button {
  margin-top: 0.5rem;
}

.c5:active {
  background-color: #0B38D9;
}

.c5:focus {
  box-shadow: 0 0 0 0.25rem #DBE3FE;
}

.c5:hover:not(:active) {
  background-color: #2852EB;
}

.c5[disabled] {
  background-color: #D9DCE9;
}

.c7 {
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: inline-grid;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
}

.c6 {
  background-color: transparent;
  border: none;
  color: #313440;
  height: auto;
  min-width: auto;
  padding: 0;
}

.c6:active {
  background-color: #DBE3FE;
}

.c6:focus {
  box-shadow: 0 0 0 0.25rem #DBE3FE;
}

.c6:hover:not(:active) {
  background-color: #F0F3FF;
}

.c6[disabled] {
  color: #D9DCE9;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  left: 0;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 1040;
  background: rgba(49,52,64,0.7);
}

.c3 {
  background: #FFFFFF;
  box-sizing: border-box;
  position: fixed;
  z-index: 1050;
  height: 100%;
  width: 100%;
}

.c4 {
  position: absolute;
  top: 0.25rem;
  right: 0.25rem;
}

.c9 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  padding: 0 1rem;
  overflow-y: auto;
}

@media (min-width:720px) {
  .c5 + .bd-button {
    margin-top: 0;
    margin-left: 0.5rem;
  }
}

@media (min-width:720px) {
  .c5 {
    width: auto;
    padding: 0;
    min-width: 2.25rem;
  }
}

@media (min-width:720px) {
  .c3 {
    border-radius: 0.25rem;
    box-shadow: 0px 2px 12px rgba(49,52,64,0.2);
    height: auto;
    max-height: 90vh;
    max-width: 720px;
  }
}

@media (min-width:1025px) {
  .c3 {
    max-height: 80vh;
  }
}

@media (min-width:720px) {
  .c4 {
    display: none;
  }
}

@media (min-width:720px) {
  .c9 {
    padding: 0 1.5rem;
  }
}

<body
  style="overflow-y: hidden;"
>
  <div />
  <div>
    <div
      aria-modal="true"
      class="c0"
      role="dialog"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r0:"
        class="c1 c2 c3"
      >
        <div
          class="c4"
        >
          <button
            class="c5 c6"
          >
            <span
              class="c7"
            >
              <svg
                aria-labelledby=":r1:"
                class="c8"
                fill="currentColor"
                height="24"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="24"
              >
                <title
                  id=":r1:"
                >
                  Close
                </title>
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M18.3 5.71a.996.996 0 0 0-1.41 0L12 10.59 7.11 5.7A.996.996 0 1 0 5.7 7.11L10.59 12 5.7 16.89a.996.996 0 1 0 1.41 1.41L12 13.41l4.89 4.89a.996.996 0 1 0 1.41-1.41L13.41 12l4.89-4.89c.38-.38.38-1.02 0-1.4"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="c9"
        >
          This is a modal
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`render open modal without backdrop 1`] = `
.c8 {
  vertical-align: middle;
  height: 1.5rem;
  width: 1.5rem;
}

.c1 {
  box-sizing: border-box;
}

.c2 {
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  -webkit-transition: all 150ms ease-out;
  transition: all 150ms ease-out;
  -webkit-transition-property: background-color,border-color,box-shadow,color;
  transition-property: background-color,border-color,box-shadow,color;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 1px solid #D9DCE9;
  border-radius: 0.25rem;
  color: #FFFFFF;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  font-size: 1rem;
  font-weight: 400;
  height: 2.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 2rem;
  outline: none;
  padding: 0 1rem;
  position: relative;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  width: 100%;
  background-color: #3C64F4;
  border-color: #3C64F4;
  font-weight: 600;
}

.c5:focus {
  outline: none;
}

.c5[disabled] {
  border-color: #D9DCE9;
  pointer-events: none;
}

.c5 + .bd-button {
  margin-top: 0.5rem;
}

.c5:active {
  background-color: #0B38D9;
}

.c5:focus {
  box-shadow: 0 0 0 0.25rem #DBE3FE;
}

.c5:hover:not(:active) {
  background-color: #2852EB;
}

.c5[disabled] {
  background-color: #D9DCE9;
}

.c7 {
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: inline-grid;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
}

.c6 {
  background-color: transparent;
  border: none;
  color: #313440;
  height: auto;
  min-width: auto;
  padding: 0;
}

.c6:active {
  background-color: #DBE3FE;
}

.c6:focus {
  box-shadow: 0 0 0 0.25rem #DBE3FE;
}

.c6:hover:not(:active) {
  background-color: #F0F3FF;
}

.c6[disabled] {
  color: #D9DCE9;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  left: 0;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 1040;
}

.c3 {
  background: #FFFFFF;
  box-sizing: border-box;
  position: fixed;
  z-index: 1050;
  height: 100%;
  width: 100%;
}

.c4 {
  position: absolute;
  top: 0.25rem;
  right: 0.25rem;
}

.c9 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  padding: 0 1rem;
  overflow-y: auto;
}

@media (min-width:720px) {
  .c5 + .bd-button {
    margin-top: 0;
    margin-left: 0.5rem;
  }
}

@media (min-width:720px) {
  .c5 {
    width: auto;
    padding: 0;
    min-width: 2.25rem;
  }
}

@media (min-width:720px) {
  .c3 {
    border-radius: 0.25rem;
    box-shadow: 0px 2px 12px rgba(49,52,64,0.2);
    height: auto;
    max-height: 90vh;
    max-width: 720px;
  }
}

@media (min-width:1025px) {
  .c3 {
    max-height: 80vh;
  }
}

@media (min-width:720px) {
  .c4 {
    display: none;
  }
}

@media (min-width:720px) {
  .c9 {
    padding: 0 1.5rem;
  }
}

<body
  style="overflow-y: hidden;"
>
  <div />
  <div>
    <div
      aria-modal="true"
      class="c0"
      role="dialog"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r2:"
        class="c1 c2 c3"
      >
        <div
          class="c4"
        >
          <button
            class="c5 c6"
          >
            <span
              class="c7"
            >
              <svg
                aria-labelledby=":r3:"
                class="c8"
                fill="currentColor"
                height="24"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="24"
              >
                <title
                  id=":r3:"
                >
                  Close
                </title>
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M18.3 5.71a.996.996 0 0 0-1.41 0L12 10.59 7.11 5.7A.996.996 0 1 0 5.7 7.11L10.59 12 5.7 16.89a.996.996 0 1 0 1.41 1.41L12 13.41l4.89 4.89a.996.996 0 1 0 1.41-1.41L13.41 12l4.89-4.89c.38-.38.38-1.02 0-1.4"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="c9"
        >
          This is a modal
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`renders destructive action button 1`] = `
.c0 {
  -webkit-transition: all 150ms ease-out;
  transition: all 150ms ease-out;
  -webkit-transition-property: background-color,border-color,box-shadow,color;
  transition-property: background-color,border-color,box-shadow,color;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 1px solid #D9DCE9;
  border-radius: 0.25rem;
  color: #FFFFFF;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  font-size: 1rem;
  font-weight: 400;
  height: 2.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 2rem;
  outline: none;
  padding: 0 1rem;
  position: relative;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  width: 100%;
  background-color: #DB3643;
  border-color: #DB3643;
  font-weight: 600;
}

.c0:focus {
  outline: none;
}

.c0[disabled] {
  border-color: #D9DCE9;
  pointer-events: none;
}

.c0 + .bd-button {
  margin-top: 0.5rem;
}

.c0:active {
  background-color: #AD0000;
}

.c0:focus {
  box-shadow: 0 0 0 0.25rem #FEDBDE;
}

.c0:hover:not(:active) {
  background-color: #CC1F1F;
}

.c0[disabled] {
  background-color: #D9DCE9;
}

.c1 {
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: inline-grid;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
}

@media (min-width:0px) {

}

@media (min-width:720px) {

}

@media (min-width:720px) {

}

@media (min-width:720px) {

}

@media (min-width:720px) {
  .c0 + .bd-button {
    margin-top: 0;
    margin-left: 0.5rem;
  }
}

@media (min-width:720px) {
  .c0 {
    width: auto;
  }
}

@media (min-width:720px) {

}

@media (min-width:1025px) {

}

@media (min-width:720px) {

}

@media (min-width:720px) {

}

@media (min-width:720px) {

}

<button
  class="c0 bd-button"
>
  <span
    class="c1"
  >
    Apply
  </span>
</button>
`;

exports[`renders secondary action button 1`] = `
.c0 {
  -webkit-transition: all 150ms ease-out;
  transition: all 150ms ease-out;
  -webkit-transition-property: background-color,border-color,box-shadow,color;
  transition-property: background-color,border-color,box-shadow,color;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 1px solid #D9DCE9;
  border-radius: 0.25rem;
  color: #FFFFFF;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  font-size: 1rem;
  font-weight: 400;
  height: 2.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 2rem;
  outline: none;
  padding: 0 1rem;
  position: relative;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  width: 100%;
  background-color: transparent;
  border-color: #3C64F4;
  color: #3C64F4;
}

.c0:focus {
  outline: none;
}

.c0[disabled] {
  border-color: #D9DCE9;
  pointer-events: none;
}

.c0 + .bd-button {
  margin-top: 0.5rem;
}

.c0:active {
  background-color: #DBE3FE;
}

.c0:focus {
  box-shadow: 0 0 0 0.25rem #DBE3FE;
}

.c0:hover:not(:active) {
  background-color: #F0F3FF;
}

.c0[disabled] {
  color: #D9DCE9;
}

.c1 {
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: inline-grid;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
}

@media (min-width:0px) {

}

@media (min-width:720px) {

}

@media (min-width:720px) {

}

@media (min-width:720px) {

}

@media (min-width:720px) {
  .c0 + .bd-button {
    margin-top: 0;
    margin-left: 0.5rem;
  }
}

@media (min-width:720px) {
  .c0 {
    width: auto;
  }
}

@media (min-width:720px) {

}

@media (min-width:1025px) {

}

@media (min-width:720px) {

}

@media (min-width:720px) {

}

@media (min-width:720px) {

}

<button
  class="c0 bd-button"
>
  <span
    class="c1"
  >
    Apply
  </span>
</button>
`;
